posthog/async_migrations/migrations/0007_persons_and_groups_on_events_backfill.py

```diff
@@ -321,9 +325,15 @@ def _check_groups_data(self, threshold=DEFAULT_ACCEPTED_INCONSISTENT_DATA_RATIO)
             ) / count() FROM events
             SAMPLE 10000000
             {where_clause}
-            """,
-            where_clause_params,
-        )[0][0]
+        """
+
+        # if CLICKHOUSE_ALLOW_PER_SHARD_EXECUTION=True, we need to ensure we run this query on a node that has the groups dictionary set up
+        if settings.CLICKHOUSE_ALLOW_PER_SHARD_EXECUTION:
+            _, _, connection_pool = get_all_shard_connections().__next__()
```

Comment:
Usually you want to avoid __ methods at all costs.

For generators, `next()` method is what you're looking for.

I consider this to be quite messy since we're now using a previously internal method but I also don't have a better suggestion for you right now.

As an aside: I'm not sure I'm actually sold on post-validation checks long-term but that's a discussion for when we do the next migration :)

Reply:
+1 on `next`

As for postcheck validations - I think usefulness can be determined on a case by case basis. For this migration particularly, it's already helped us a good amount. Both gave me confidence when self-hosted users reported back that things ran successfully but also already helped us catch data inconsistencies on a user instance